netplugin/netd.go

```diff
@@ -464,28 +465,37 @@ func handleStateEvents(netPlugin *plugin.NetPlugin, crt *crt.CRT, opts cliOpts,
 	go handleEndpointEvents(netPlugin, crt, opts, retErr)
 }
 
-func handleEvents(netPlugin *plugin.NetPlugin, crt *crt.CRT, opts cliOpts) error {
-
-	recvErr := make(chan error, 1)
+func handleEvents(netPlugin *plugin.NetPlugin, crt *crt.CRT, opts cliOpts) {
+	recvErr := make(chan error)
+	recvEventErr := make(chan error)
 
 	//monitor and process state change events
 	go handleStateEvents(netPlugin, crt, opts, recvErr)
+	startDockerEventPoll(netPlugin, crt, recvEventErr, opts)
+
+	go func() {
+		for {
+			err := <-recvEventErr
+			if err != nil {
+				log.Warnf("Failure occured talking to docker events. Sleeping for a second and retrying. Error: %s", err)
+			}
+
+			time.Sleep(1 * time.Second)
+			// FIXME note that we are assuming this goroutine crashed, which if not
+			// might generate double events. Synchronize this with a channel.
+			startDockerEventPoll(netPlugin, crt, recvErr, opts)
+		}
+	}()
+
+}
 
+func startDockerEventPoll(netPlugin *plugin.NetPlugin, crt *crt.CRT, recvErr chan error, opts cliOpts) {
 	if !opts.nativeInteg {
 		// start docker client and handle docker events
 		// wait on error chan for problems handling the docker events
 		dockerCRT := crt.ContainerIf.(*docker.Docker)
-		dockerCRT.Client.StartMonitorEvents(handleDockerEvents, recvErr,
-			netPlugin, crt)
-	}
-
-	err := <-recvErr
```

Comment:
now with error-channels separated out, we can bring back this code to block on `recvErr` channel inside `handleEvents()` to correctly signal errors to the calling function.